src/java.base/share/classes/jdk/internal/module/ModuleInfoExtender.java

```diff
@@ -148,90 +150,53 @@ public void write(OutputStream out) throws IOException {
-                        }
-                    }
-                    public void visitPackage(String existingPackage) {
-                        // skip packages if there is a new set of packages
-                        if (packages == null) {
-                            super.visitPackage(existingPackage);
-                        }
-                    }
-                };
+        var cm = Classfile.parse(in.readAllBytes());
+        Version v = ModuleInfoExtender.this.version;
+        return cm.transform(ClassTransform.endHandler(clb -> {
+            // ModuleMainClass attribute
+            if (mainClass != null) {
+                clb.with(ModuleMainClassAttribute.of(ClassDesc.of(mainClass)));
             }
-            @Override
-            public void visitAttribute(Attribute attr) {
-                String name = attr.type;
-                // drop existing attributes if there are replacements
-                if (name.equals(ClassFileConstants.MODULE_TARGET)
-                    && targetPlatform != null)
-                    return;
-                if (name.equals(ClassFileConstants.MODULE_RESOLUTION)
-                    && moduleResolution != null)
-                    return;
-                if (name.equals(ClassFileConstants.MODULE_HASHES)
-                    && hashes != null)
-                    return;
-
-                super.visitAttribute(attr);
 
+            // ModulePackages attribute
+            if (packages != null) {
+                List<PackageDesc> packageNames = packages.stream()
+                        .sorted()
+                        .map(s -> PackageDesc.of(s))
+                        .toList();
+                clb.with(ModulePackagesAttribute.ofNames(packageNames));
             }
-        };
-
-        List<Attribute> attrs = new ArrayList<>();
-        attrs.add(new ModuleTargetAttribute());
-        attrs.add(new ModuleResolutionAttribute());
-        attrs.add(new ModuleHashesAttribute());
-        cr.accept(cv, attrs.toArray(new Attribute[0]), 0);
-
-        // add ModuleTarget, ModuleResolution and ModuleHashes attributes
-        if (targetPlatform != null) {
-            cw.visitAttribute(new ModuleTargetAttribute(targetPlatform));
-        }
-        if (moduleResolution != null) {
-            int flags = moduleResolution.value();
-            cw.visitAttribute(new ModuleResolutionAttribute(flags));
-        }
-        if (hashes != null) {
-            String algorithm = hashes.algorithm();
-            List<String> names = new ArrayList<>();
-            List<byte[]> values = new ArrayList<>();
-            for (String name : hashes.names()) {
-                names.add(name);
-                values.add(hashes.hashFor(name));
-            }
-            cw.visitAttribute(new ModuleHashesAttribute(algorithm, names, values));
-        }
 
-        return cw.toByteArray();
+            // add ModuleTarget, ModuleResolution and ModuleHashes attributes
```

Comment:
This adds or replaces if the attribute exists.  Perhaps just drop "add" from the comment?

Reply:
fixed, thanks.